Add proposed 'window/progress' notification

Xanewok opened the pull request:
Closes #43.

Sorry, I'm too itchy to not open it - the original proposed PR has been opened 1.5 years ago so it's about time :grinning: 

This adds `proposed` feature which enables proposed LSP extensions.

RLS PR that's working and which only reexports the types - https://github.com/Xanewok/rls/commit/68da7677c6774ffb758a1a609a133ac5711a24e6

Xanewok commented on src/notification.rs:
```diff
@@ -59,6 +59,10 @@ macro_rules! lsp_notification {
         $crate::notification::DidChangeWorkspaceFolders
     };
+    // Proposed
+    // ("window/progress") => {
```
Initially I wanted to warn the user if they're trying to use a feature that's proposed but since the macro has to return specifically a type, I wasn't sure what to return here, so I just commented it out

Marwes replied:
Should be fine to just return the type as normal, if `proposed` is not enabled then it will just be a compile error. (Add a comment on it so people can at least read to source to see why it does not work).

Xanewok commented on src/notification.rs:
```diff
@@ -252,6 +256,24 @@ impl Notification for PublishDiagnostics {
+    /// The progress notification is sent from the server to the client to ask
+    /// the client to indicate progress.
+    #[derive(Debug)]
+    pub enum Progress {}
```
it's weird, although I'm doing `pub use proposed::*;` in both `src/lib.rs` and `src/notification.rs` I can refer to
```rust
lsp_types::ProgressParams
```
but not
```rust
lsp_types::notification::Progress // needs `proposed::Progress`
```
any ideas why/how to solve that? Do we want to retain the explicit `proposed` mod?

Marwes replied:
No clue.

Would be nice to have it consistently exported though. If the module needs to be removed for that then removing it seems like the better option.

Xanewok replied:
Removed the module then

Marwes commented on src/notification.rs:
```diff
@@ -290,5 +312,8 @@ mod test {
         check_macro!("workspace/didChangeWorkspaceFolders");
+
+        // Proposed
+        // check_macro!("window/progress");
```
```suggestion
        #[cfg(feature = "supported")]
        check_macro!("window/progress");
```

And try compiling with and without `supported` on CI.

Marwes commented on Cargo.toml:
```diff
@@ -24,3 +24,8 @@ serde_json = "1.0.0"
+default = []
+# Enables proposed LSP extensions.
+# NOTE: No semver compatibility is guaranteed for types enabled by this feature.
+proposed = []
```
Nit: Newline

Marwes commented:
What is left after https://github.com/Microsoft/vscode-languageserver-node/pull/261 to make it properly part of the spec?

Xanewok commented:
Done, addressed the feedback. While modifying .travis.yml it made sense to test also `stable` since this doesn't require any nightly features - should I remove it from the matrix?

> What is left after Microsoft/vscode-languageserver-node#261 to make it properly part of the spec?

That's... actually a good question. So the PR is proposed and a reference implementation is provided as per their recommendation. I imagine the proposed features will make it in next LSP version (3.15?) so that the users can start using it.

Xanewok commented:
The checks are all green now. I also added stable support (since we're not using any nightly features) - is that okay?

Marwes commented:
Released as 0.57.

Xanewok commented:
Thanks! :tada:

Xanewok commented:
For better or worse, this was further split into 4 different message literals - https://github.com/Microsoft/language-server-protocol/issues/70#issuecomment-483252666

Marwes commented:
Whelp, I guess this still works until it gets properly stabilized though?

Xanewok commented:
Yeah, it does; this was a proposed extension by Facebook et al. and couple of servers adopted this `window/progress` version, so it's useful on its own. When the current version is stabilized we should update the types here.
